File: crates/test_host/src/widgets/combo_box.rs
use crate::{
    bindings::iced::app::combo_box::ComboBox,
    widgets::{ToElement, WrapperRenderer, WrapperTheme},
};

use super::Message;

impl ToElement for ComboBox {
    fn to_element<'a, Theme, Renderer>(
        self,
        _resource_table: &mut wasmtime::component::ResourceTable,
    ) -> iced::Element<'a, Message, Theme, Renderer>
    where
        Theme: WrapperTheme + 'a,
        Renderer: WrapperRenderer + 'a,
    {
        let ComboBox {
            options,
            placeholder,
            selected,
            on_selected,
            on_input,
            on_option_hovered,
            on_open,
            on_close,
            padding,
            size,
            line_height,
            width,
        } = self;

        let state: &'static iced::widget::combo_box::State<String> =
            Box::leak(Box::new(iced::widget::combo_box::State::new(options)));

        let mut combo_box =
            iced::widget::ComboBox::new(state, &placeholder, selected.as_ref(), move |v| Message {
                id: on_selected,
                content: crate::bindings::Message::StringType(v),
            });

        if let Some(id) = on_input {
            combo_box = combo_box.on_input(move |v| Message {
                id,
                content: crate::bindings::Message::StringType(v),
            });
        }

        if let Some(id) = on_option_hovered {
            combo_box = combo_box.on_option_hovered(move |v| Message {
                id,
                content: crate::bindings::Message::StringType(v),
            });
        }

        if let Some(id) = on_open {
            combo_box = combo_box.on_open(id.into());
        }

        if let Some(id) = on_close {
            combo_box = combo_box.on_close(id.into());
        }

        if let Some(padding) = padding {
            combo_box = combo_box.padding(padding);
        }

        if let Some(size) = size {
            combo_box = combo_box.size(size);
        }

        if let Some(line_height) = line_height {
            combo_box = combo_box.line_height(line_height);
        }

        if let Some(width) = width {
            combo_box = combo_box.width(width);
        }

        combo_box.into()
    }
}

File: packages/igloo-ts/src/widgets/combo-box.ts
import { ComboBox as WitComboBox } from 'iced:app/combo-box@0.1.0';
import type { CallbackId } from 'iced:app/callbacks@0.1.0';
import type { Length } from 'iced:app/length@0.1.0';
import type { Padding } from 'iced:app/padding@0.1.0';
import type { LineHeight } from 'iced:app/text@0.1.0';
import { Element, type IntoElement } from '../element.js';
import { pushFixed, pushString } from '../callbacks.js';

/**
 * Builder for creating ComboBox widgets.
 * A ComboBox allows selection from a list of options with text input filtering.
 *
 * @example
 * ```typescript
 * const comboBox = ComboBox.new(
 *   ['Apple', 'Banana', 'Cherry'],
 *   'Search fruits...',
 *   (fruit) => ({ type: 'fruitSelected', fruit }),
 *   state.selectedFruit,
 * );
 * ```
 *
 * @typeParam Msg - The message type this combo box emits; inferred from its handlers.
 */
export class ComboBox<Msg = never> implements IntoElement<Msg> {
  private raw: WitComboBox;

  private constructor(
    options: string[],
    placeholder: string,
    selected: string | undefined,
    onSelected: CallbackId,
  ) {
    this.raw = new WitComboBox(options, placeholder, selected, onSelected);
  }

  /**
   * Create a new ComboBox builder.
   * @param onSelected - Handler called with the selected option
   * @param selected - The currently selected option, if any
   */
  static new<const M>(
    options: string[],
    placeholder: string,
    onSelected: (value: string) => M,
    selected?: string,
  ): ComboBox<M> {
    return new ComboBox(options, placeholder, selected, pushString(onSelected));
  }

  /** Set the message to emit when text is input (for filtering) */
  onInput<const M>(mapper: (value: string) => M): ComboBox<Msg | M> {
    this.raw.onInput(pushString(mapper));
    return this as unknown as ComboBox<Msg | M>;
  }

  /** Set the message to emit when an option is hovered */
  onOptionHovered<const M>(mapper: (value: string) => M): ComboBox<Msg | M> {
    this.raw.onOptionHovered(pushString(mapper));
    return this as unknown as ComboBox<Msg | M>;
  }

  /** Set the message to emit when the combo box is opened */
  onOpen<const M>(msg: () => M): ComboBox<Msg | M> {
    this.raw.onOpen(pushFixed(msg()));
    return this as unknown as ComboBox<Msg | M>;
  }

  /** Set the message to emit when the combo box is closed */
  onClose<const M>(msg: () => M): ComboBox<Msg | M> {
    this.raw.onClose(pushFixed(msg()));
    return this as unknown as ComboBox<Msg | M>;
  }

  /** Set the padding */
  padding(padding: Padding): this {
    this.raw.padding(padding);
    return this;
  }

  /** Set the text size */
  size(size: number): this {
    this.raw.size(size);
    return this;
  }

  /** Set the line height */
  lineHeight(lineHeight: LineHeight): this {
    this.raw.lineHeight(lineHeight);
    return this;
  }

  /** Set the width */
  width(width: Length): this {
    this.raw.width(width);
    return this;
  }

  /** Convert to Element */
  intoElement(): Element<Msg> {
    return new Element(WitComboBox.intoElement(this.raw));
  }
}
